Add 100 node scale test workflow

learnitall opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [X] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [X] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Provide a title or release-note blurb suitable for the release notes.
- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [X] Thanks for contributing!

This PR culminates in adding a workflow for running a ClusterLoader2 load test in a 100-node cluster. It builds on the initial workflow added in https://github.com/cilium/cilium/pull/28362. 

This PR also contains work that modularizes common steps between scale test workflows into separate actions. 

Co-authored-by: @marseel 

```release-note
Add 100 node scale test workflow
```

marseel committed:
ci: Modularize scale-test workflow

This commit modularizes steps in the scale test workflow by
turning them into their own actions. These actions can be used
by new scale test workflows in the future to reduce code
duplication.

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>

marseel committed:
ci: Add 100 nodes scale test

This commit adds a 100 node scale test workflow, which creates a 100
node cluster and runs a full CL2 test suite. Three changes were needed
to the current scale test actions:

1. Adjust the create-cluster action to use a larger network. Kops
   will provision the 100 node cluster in a /20 by default, however
   this doesn't leave enough address space.
2. Add a new action to create an additional instance group within
   the cluster. This action is used to deploy a larger node inside
   the 100 node cluster which CL2 will use to host Prometheus.
3. Bump the timeout used for the validate-cluster action from 20m to 45m

Signed-off-by: Marcel Zieba <marcel.zieba@isovalent.com>
Co-authored-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ci: Add CLUSTER_NAME variable in vars step in scale test workflows

This commit adds a dedicated variable for the target cluster's name,
meaning steps in the workflow do not have to construct the name from
other variables.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ci: Change filename of scale-test-gce workflow to be more specific

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ci: Create separate action for adjusting kops firewall rules

When creating a cluster using kops for scale tests run with CL2, a
manual modification is made to a firewall rule that kops deploys to
manage traffic sent from workers to control plane nodes. This manual
modification is reset every time a call to `kops update` is made. This
commit creates a separate action for this step so it can be called at
the appropriate time in scale test workflows.

See: kubernetes/perf-tests#2319

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ci: Update CL2 in 100 node scale test workflow

This commit updates the commit SHA for ClusterLoader2 in the 100 node
scale test workflow, which pulls in recent changes to allow the 100 node
scale test workflow to successfully pass.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ci: Adjust schedule of 100 node scale test to run every week day

This commit changes the cron schedule of the 100 node scale test to run
every business day. After two weeks, this commit will be reverted. The
goal is to collect lots of data for the next two weeks and ensure the
workflow is working properly.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

marseel commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -0,0 +1,222 @@
+            --provider=gce \
+            --enable-prometheus-server \
+            --tear-down-prometheus-server=false \
+            --nodes=10 \
```
We should set it to 100.

learnitall replied:
Oh nice catch, I missed this, thank you!

brlbil commented on .github/actions/scale-tests/cleanup-cluster/action.yaml:
```diff
@@ -0,0 +1,19 @@
+          ./kops delete cluster \
+            ${{ inputs.cluster_name }} \
+            --state ${{ inputs.kops_state }} \
+            --yes
```
It is a nit, a new line end of the file is missing. There are more please see below.

brlbil commented on .github/actions/scale-tests/create-cluster/action.yaml:
```diff
@@ -0,0 +1,82 @@
+    - name: Wait for cluster control plane
+      shell: bash
+      run: |
+        until kubectl get nodes; do sleep 45s; done
```
New line

brlbil commented on .github/actions/scale-tests/export-results/action.yaml:
```diff
@@ -0,0 +1,28 @@
+        BUILD_ID="$(date +%s)"
+        EXPORT_DIR="${{ inputs.results_bucket }}/logs/${{ inputs.test_name }}-${{ github.ref_name }}/${BUILD_ID}"
+        gsutil -m cp -R ${{ inputs.artifacts }} $EXPORT_DIR/artifacts/
+        gsutil -m cp sha.txt run-url.txt ${{ inputs.other_files }} $EXPORT_DIR/
```
New line

brlbil commented on .github/actions/scale-tests/setup-firewall/action.yaml:
```diff
@@ -0,0 +1,23 @@
+        # Rules could be added individually, but this is a bit simpler,
+        # matches the rules used by kops to manage cp-to-node and node-to-node
+        # traffic, and is an acceptable risk since the cluster is ephemeral and not
+        # running in prod.
```
New line

brlbil commented on .github/actions/scale-tests/validate-cluster/action.yaml:
```diff
@@ -0,0 +1,20 @@
+          ${{ inputs.cluster_name }} \
+          --state ${{ inputs.kops_state }} \
+          --wait 45m \
+          --count 3
```
New line

brlbil commented on .github/workflows/scale-test-node-throughput-gce.yaml:
```diff
@@ -1,8 +1,8 @@
-name: Scale Test GCE
+name: Node throughput test
```
It is a nit, can we change it to 'Node Throughput Test'

brlbil commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -0,0 +1,222 @@
+name: 100 Nodes scale test
```
For consistency, can we make it title 'Nodes Scale Test'
Also a question, it seems a very spesific name '100 Nodes Scale Test', will there be other tests like 250?

marseel replied:
"100" is on purpose here. The idea is that we can compare easily 1:1 the results of this scale test with [k8s scale test](https://testgrid.k8s.io/sig-scalability-gce#gce-cos-master-scalability-100) to see in what areas we under/overperform. Next one would be 5k nodes :)

marseel commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -0,0 +1,222 @@
+            --experimental-prometheus-snapshot-to-report-dir=true \
+            --kubeconfig=$HOME/.kube/config \
+            --testoverrides=./testing/overrides/load_throughput.yaml \
+            --testoverrides=./testing/experiments/use_simple_latency_query.yaml \
```
I must have missed it when adding tee. Test https://github.com/cilium/cilium/actions/runs/6896887510/job/18763936013 failed with error:
```
measurement call SchedulingThroughput - SchedulingThroughput error: scheduler throughput: actual throughput 21.000000 lower than threshold 100.000000
```
This is fine, as 100 nodes cluster is expected to have ~20 throughput, but the test itself didn't fail, because of tee. 
We should set `set -o pipefail` before and also turn off scheduling throughput for this test with `CL2_SCHEDULER_THROUGHPUT_THRESHOLD=0` (this is what k8s OSS test does for 100 nodes test [here](https://github.com/kubernetes/test-infra/blob/a6a331d32674564c9b70da46127c77279c3081e5/config/jobs/kubernetes/sig-scalability/sig-scalability-release-blocking-jobs.yaml#L210) )

learnitall replied:
I set the `shell: bash` parameter, which will set pipefail: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell

marseel commented on .github/actions/scale-tests/create-ig/action.yaml:
```diff
@@ -0,0 +1,60 @@
+          # and guaranteed to be what we set it to.
+          # Otherwise, the value given will be appended to the zone list.
+          ./kops edit ig ${{ inputs.ig_name }} \
+            --state ${{ inputs.kops_state }} \
```
This is so much better than what I did before, thanks!

marseel reviewed (approved):
Thanks! Looks good, just one comment :)

aanm commented on .github/actions/scale-tests/cleanup-cluster/action.yaml:
```diff
@@ -0,0 +1,19 @@
+name: 'Cleanup kops cluster'
```
Perhaps it would make sense to move the scale-tests GH action into a dedicated repository? Similar to what we have for the [little-vm-helper-action](https://github.com/cilium/little-vm-helper)

learnitall replied:
I can see that. I think moving to the scaffolding repository could make sense, but my only hesitation is that the scale test actions are specific to the cilium repo right now and may not be used anywhere else for a while.

Could we start by leaving them here and then move them as a follow-up?

aanm replied:
> I can see that. I think moving to the scaffolding repository could make sense, but my only hesitation is that the scale test actions are specific to the cilium repo right now and may not be used anywhere else for a while.

Don't worry about it. We have other repositories that are only specific to the cilium repo and do not belong to this repository. I've created https://github.com/cilium/scale-tests-action for this purpose.

learnitall replied:
Done. Thank you for your help!

aanm commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -0,0 +1,222 @@
+      - name: Clone ClusterLoader2
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        with:
+          repository: learnitall/perf-tests
```
why your fork? We can have forks in cilium organization until the upstream repository has the patch.

learnitall replied:
I used my fork just for convenience, and I didn't think moving to the Cilium org was a blocker. We can move it there if you'd like, but I'd prefer if we merge the workflow first so we can get some runs going and then work on migrating to the Cilium org. How does that sound?

aanm replied:
It is easier for others to contribute if it's part of the cilium org. I've created https://github.com/cilium/perf-tests

learnitall replied:
I updated the workflows to point to the fork

aanm commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -0,0 +1,222 @@
+          kops_state: ${{ secrets.GCP_PERF_KOPS_STATE_STORE }}
+          project_id: ${{ secrets.GCP_PERF_PROJECT_ID }}
+
+      - name: Create IG for resource heavy deployments
```
Can we rename `IG` to  `Instance Group`? It's not clear to everyone what `ig` means.

learnitall replied:
Done

aanm commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -0,0 +1,222 @@
+      - name: Setup firewall rules
+        uses: ./.github/actions/scale-tests/setup-firewall
+        with:
+          cluster_name: ${{ steps.vars.outputs.cluster_name }}
```
This step could be a parameter (`setup-firewall: bool`) of the `create-ig` which would then setup the firewall accordingly after cluster creation.

learnitall replied:
The setup firewall action is also used as a step in the node throughput test, which does not call the `create-ig` action. Can we keep the firewall step separate?

aanm replied:
> The setup firewall action is also used as a step in the node throughput test, which does not call the `create-ig` action. Can we keep the firewall step separate?

I see, and is it necessary that `create-instance-group` is a separate step? These are internal details of the scale tests, they are not relevant for the users of this GH action.

learnitall replied:
IMO I think having the `create-instance-group` as a separate step can be useful. We could just take the content of the step and put it into the 100 node scale test workflow, but creating an instance group is something that we'll have to do anyways in other scale tests we want to implement in the future. Creating another instance group to host specific resources is a common pattern in perf/scale tests.

aanm commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -0,0 +1,222 @@
+
+      - name: Wait for cluster to be ready
+        uses: ./.github/actions/scale-tests/validate-cluster
+        timeout-minutes: 45
```
45 minutes seems to high, do we need to be 45 mins?

learnitall replied:
No it can be lowered, good catch.

aanm commented on .github/workflows/scale-test-100-gce.yaml:
```diff
@@ -0,0 +1,222 @@
+      - name: Run CL2
+        id: run-cl2
+        working-directory: ./perf-tests/clusterloader2
+        run: |
```
This one should have a timeout instead.

learnitall replied:
Added

learnitall committed:
ci: Fix capitalization of node throughput workflow name

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ci: Add timeout to CL2 step in node throughput scale test workflow

This commit adds an explicit timeout to the step in the node throughput
scale test workflow that runs CL2.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ci: Move scale-test workflows and perf-tests fork to cilium org

All scale-test related actions, meaning, everything under
.github/actions/scale-tests, has been moved to the
cilium/scale-tests-action repository. The fork of kubernetes/perf-tests
has been moved from learnitall/perf-tests to cilium/perf-tests.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ci: Use env var instead of secret for scale tests results bucket

This bucket is exposed to the internet and meant to be accessed by the
community, therefore the bucket name shouldn't be a secret.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

learnitall committed:
ci: Specify bash as shell for CL2 step in node throughput test

This commit modifies the CL2 step in the node throughput scale test to
specify 'bash' as the shell, which toggles the pipefail option. This is
important, because without the pipefail option, this step would always
succeed.

Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>
